File: demo/ruby/broadcast.rb
#!/usr/bin/ruby

# A basic broadcast system. Keeps track of a set of messages, inserted via
# `broadcast` requests from clients.

require_relative 'node.rb'
require 'set'

class BroadcastNode
  def initialize
    @node = Node.new
    @neighbors = []
    @lock = Mutex.new
    @messages = Set.new

    @node.on "topology" do |msg|
      @neighbors = msg[:body][:topology][@node.node_id.to_sym]
      @node.log "My neighbors are #{@neighbors.inspect}"
      @node.reply! msg, {type: "topology_ok"}
    end

    @node.on "read" do |msg|
      @lock.synchronize do
        @node.reply! msg, {type: "read_ok",
                           messages: @messages.to_a}
      end
    end

    @node.on "broadcast" do |msg|
      m = msg[:body][:message]
      @lock.synchronize do
        unless @messages.include? m
          @messages.add m
          @node.log "messages now #{@messages}"

          # Broadcast this message to neighbors (except whoever sent it to us)
          @node.other_node_ids.each do |neighbor|
            unless neighbor == msg[:src]
              @node.rpc! neighbor, {type: "broadcast", message: m} do |res|
                # Eh, whatever
              end
            end
          end
        end
      end
      @node.reply! msg, {type: "broadcast_ok"}
    end
  end

  def main!
    @node.main!
  end
end

BroadcastNode.new.main!

File: demo/python/broadcast.py
#!/usr/bin/env python3

import asyncio
from maelstrom import Node, Body, Request

node = Node()

messages: set[int] = set()
cond = asyncio.Condition()


async def add_messages(values: set[int]) -> None:
    if values <= messages:
        return
    messages.update(values)
    async with cond:
        cond.notify_all()


@node.handler
async def broadcast(req: Request) -> Body:
    msg = req.body["message"]
    await add_messages({msg})
    return {"type": "broadcast_ok"}


@node.handler
async def broadcast_many(req: Request) -> Body:
    msgs = req.body["messages"]
    await add_messages(set(msgs))
    return {"type": "broadcast_many_ok"}


@node.handler
async def read(req: Request) -> Body:
    return {"type": "read_ok", "messages": list(messages)}


@node.handler
async def topology(req: Request) -> Body:
    neighbors = req.body["topology"][node.node_id]
    for n in neighbors:
        node.spawn(gossip_task(n))
    return {"type": "topology_ok"}


async def gossip_task(neighbor: str) -> None:
    sent = set()
    while True:
        # assert sent <= messages
        if len(sent) == len(messages):
            async with cond:
                # Wait for the next update to our message set.
                await cond.wait_for(lambda: len(sent) != len(messages))

        to_send = messages - sent
        body = {"type": "broadcast_many", "messages": list(to_send)}
        resp = await node.rpc(neighbor, body)
        if resp["type"] == "broadcast_many_ok":
            sent.update(to_send)


node.run()
